### ISSUE TITLE
Fused batch renormalization

### ISSUE THREAD

**Issue by ygoncharov** [tensorflow/tensorflow#9141]
Recently introduced batch renormalization could be very useful, but without support for a fused batch norm, it has limited use (XLA JIT still has too many problems to be used instead of a fused version)

**Comment by jart** [comment]
@fchollet Our friend @ygoncharov has a brief request relating to batch normalization, which I noticed you worked on in the past.

**Comment by fchollet** [comment]
We will support fused BN in the future (ideally this should be done via XLA rather than via explicit TF Python code, but we will see). If you need fused BN for the time being, you can simply use the BN layer in `contrib.layers` (although it is now deprecated).

**Comment by jart** [comment]
@fchollet Is there value in keeping this issue open and assigning it to you? That way when fused BN come along, you can put `Fixed #9141` in the CL description to notify subscribers.

**Comment by ygoncharov** [comment]
Please note that my request was about batch REnormalization (https://arxiv.org/pdf/1702.03275.pdf) that recently appeared as an option for batch_norm in contrib.layers, not about regular batch norm.

>  contrib.layers (although it is now deprecated).
@fchollet is the layer lib (and the entire Slim I guess) officially depricated now? I so, we better start converting back to Keras early :)

**Comment by Caenorst** [comment]
From my understanding, Fused Batch Norm is mostly faster because of CuDNN implementation.

CuDNN implementation return the variance (while it internally compute the inverse of stddev to normalize), but Batch Renormalization apply the moving average on the the stddev, so you would have to apply again rsqrt to find the inverse of stddev and the stddev, which is I guess the slowest part of the batch norm.

So unless Nvidia decide to do a CuDNN implementation of batch renormalization, XLA should be the way to go.

**Comment by jart** [comment]
What @Caenorst said sounds like a compelling reason to hold off on implementing this feature for the time being. If my cursory assessment is inaccurate, please let me know and I'll re-open this issue.

**Comment by ygoncharov** [comment]
@Caenorst are you sure that fused batch norm is implemented on the CuDNN level, not on the tf ops level? There are cudnnBatchNormalization* functions, but I am not aware of any fused versions of those.

It seems that a new tensorflow op would be enough to implement batch renormalization with fusion, no new CuDNN functions are needed.

@jart it seems that it is a reasonable feature request, unless there is a decision to do all such optimizations via XLA JIT

**Comment by Caenorst** [comment]
Fused Batch Norm **is** CuDNN batch normalization, on the original/non-fused batch normalization code I did not found any CuDNN call.

CuDNN is not open sourced so I don't think there is a point to rebuild it from scratch while one of the main objective of XLA JIT is to "fuse" op to get speedup.

**Comment by taion** [comment]
@Caenorst @jart 

That's incorrect. The non-fused batch normalization is dramatically slower than the fused version when using NCHW – in fact, when using non-fused BN, NCHW is _slower_ than NHWC, despite all the convolutions going faster: https://github.com/tensorflow/tensorflow/issues/7551#issuecomment-280421351

Non-fused batch norm currently slows things down in NCHW that it's unusable in that mode, and is actually a pretty significant performance landmine.

**Comment by taion** [comment]
The minor performance hit from not using cuDNN is "whatever" to me, but the much larger performance hit of having to drop to NHWC to not suffer that massive performance penalty detailed in my link above is a pretty big deal.

The **6x** slowdown with non-fused batch norm in NCHW means that, in practice, this is a pretty big performance issue for most users, to the point where I think it is in fact quite compelling.

**Comment by ygoncharov** [comment]
Indeed, batch renorm in NCHW is almost broken if used without XLA JIT. It seems that the options are:
1. Implement a fused version of batch renorm as a custom op without using CuDNN (it is a bigger task than an original fused batch norm that, as @Caenorst correctly noted, relies on CuDNN)
2. Make a strategic decision to rely on XLA JIT for all such cases (even though it is still in alpha)
3. Wait for CuDNN 7+ that hopefully implements batch renorm functions and then implement a corresponding op

**Comment by Caenorst** [comment]
@taion Thank you for you correction. So looking at issue #7551 doesn't it mean that the issue is at a lower level than Batch Normalization ? Or should we build a fused kernel every time we have to do a broadcast add or scale on non-last axis (or use bias_add given your last comment) ?

**Comment by taion** [comment]
It's probable that I should split that into two issues. I don't think the broadcasted arithmetic is the only issue with non-fused BN when using NCHW. And anyway, as far as I'm aware, there's no multiplication equivalent to `bias_add`, so there's not a good way to implement the scale operation.

Certainly there is some problem here at a lower level than non-fused BN, but in practice my models ran fine as soon as I switched to the fused BN, so perhaps those effects are less important.

**Comment by jart** [comment]
@broune I want to bring to your attention a discussion relating to XLA JIT. How would you recommend this issue be triaged?

**Comment by broune** [comment]
We are aware that batch normalization can be improved on GPU (and presumably on CPU). We are currently working on enabling fusion nodes with multiple outputs (also for other reasons), and this may help here, since some batchnorm division ops are not getting fused because they are needed by several other ops, but those ops, including the divide, could be combined into a single fusion node if fusion nodes could have multiple outputs. That would prevent the intermediate data from getting written to memory. The division nodes in particular are not getting fused because the XLA heuristic is that these are expensive on GPU and so they are not duplicated into the consuming fusion nodes.

There also seems to be a problem where a broadcast op is not fused into a fusion node that does a reduce, so the broadcast ends up being written to memory and then read back, which doesn't seem good, especially since the source data for the broadcast is small. I'll pass that problem on.

We may in future consider adding a dedicated batchnorm op to XLA, since batchnorm is too large of a subgraph to reasonably pattern match, though not clear at this point if that will end up being useful.

**Comment by jart** [comment]
Thank you for the information @broune. I'll assign so you can determine how this issue should be handled from here.

**Comment by Caenorst** [comment]
Also, even for the [fused batch norm implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fused_batch_norm_op.cc#L248) they actually transform the data_format from NCHW to NHWC before applying the CuDNN batch normalization.

While bias_add can natively support NCHW and NHWC (there are both "BiasNHWCKernel" and "BiasNCHWKernel"), I have no idea if there is a speed difference between both.

I guess a kind of "broadcast_mul" in the same manner than "bias_add" would be useful, for many case (batch renorm being one of them).

**Comment by taion** [comment]
@Caenorst

That's going the other way around, no? See also https://github.com/tensorflow/tensorflow/issues/8286. For those of us using cuDNN, I don't think there's any reason why we wouldn't want to always use NCHW anyway, modulo the performance bugs we're discussing here.

**Comment by joeyearsley** [comment]
Will this ever be supported? - Using ultra large images constrain batch sizes to 1, so renormalization is a must for myself. I use `NCHW` format, hence the interest.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by broune** [comment]
I assume that you're using this on GPU. For BatchNorm, we do have an op in XLA, and lately Justin Lebar tried it on GPU, but concluded that XLA is actually faster than cuDNN for this operation, IIRC. So that may also be true for batch RE-normalization at this point. Assigning over to Justin to comment on the GPU angle on this.

**Comment by jlebar** [comment]
> For BatchNorm, we do have an op in XLA, and lately Justin Lebar tried it with the GPU backend of XLA using cuDNN, but concluded that raw XLA is actually faster than cuDNN for this operation, IIRC

Just to be clear, what I saw was that on our models, using cudnn batchnorm was roughly the same speed as our XLA implementation.  However, our lowering of these particular XLA ops to PTX is not very good at the moment -- I think it could be made significantly faster, at which point we would beat cudnn.

I don't see why batch re-normalization would be different, although someone would have to benchmark.

Note that the comparison I was making was between everything-compiled-with-xla+cudnn-batchnorm vs everything-compiled-with-xla+pure-xla-batchnorm.  If you tried instead compiling the *just* one TF batchnorm (or batch-renorm) op with XLA, then XLA wouldn't be able to fuse the op's inputs or outputs, and you might indeed find that cudnn is faster than XLA.

I'm going to unassign myself since it doesn't sound like there's work that we need to do here that's separate from the regular XLA performance work we're already doing.  (This is a good thing -- part of the promise of XLA is that we don't have to open a bug every time we have a new op that we want to optimize.)

**Comment by shivaniag** [comment]
With @jlebar's comment, I am closing this issue. If there is any further discussion on this feel free to reopen the issue..
